src/app/clusters/thermostat-server/thermostat.cpp

```diff
@@ -45,115 +45,182 @@ void emberAfThermostatClusterServerInitCallback(void)
+bool emberAfThermostatClusterGetWeeklyScheduleCallback(EndpointId aEndpointId, chip::app::CommandHandler * commandObj,
+                                                       uint8_t daysToReturn, uint8_t modeToReturn)
 {
     // TODO
     return false;
 }
 
-bool emberAfThermostatClusterSetWeeklyScheduleCallback(app::CommandHandler * commandObj, uint8_t numberOfTransitionsForSequence,
-                                                       uint8_t daysOfWeekForSequence, uint8_t modeForSequence, uint8_t * payload)
+bool emberAfThermostatClusterSetWeeklyScheduleCallback(EndpointId aEndpointId, chip::app::CommandHandler * commandObj,
+                                                       uint8_t numberOfTransitionsForSequence, uint8_t daysOfWeekForSequence,
+                                                       uint8_t modeForSequence, uint8_t * payload)
 {
     // TODO
     return false;
 }
 
-bool emberAfThermostatClusterSetpointRaiseLowerCallback(app::CommandHandler * commandObj, uint8_t mode, int8_t amount)
+int32_t EnforceHeatingSetpointLimits(int32_t HeatingSetpoint, EndpointId endpoint)
+{
+    // Optional Mfg supplied limits
+    int16_t AbsMinHeatSetpointLimit = 0x02bc; // 7C (44.5 F) is the default
+    int16_t AbsMaxHeatSetpointLimit = 0x0bb8; // 30C (86 F) is the default
+
+    // Optional User supplied limits
+    int16_t MinHeatSetpointLimit = 0x02bc; // 7C (44.5 F) is the default
+    int16_t MaxHeatSetpointLimit = 0x0bb8; // 30C (86 F) is the default
+
+    // Attempt to read the setpoint limits
+    // Absmin/max are manufacturer limits
+    // min/max are user imposed min/max
+
+    // Note that the limits are initialized above per the spec limits
+    // if they are not present emberAfReadAttribute() will not update the value so the defaults are used
+
+    emberAfReadAttribute(endpoint, ZCL_THERMOSTAT_CLUSTER_ID, ZCL_ABS_MIN_HEAT_SETPOINT_LIMIT_ATTRIBUTE_ID, CLUSTER_MASK_SERVER,
+                         (uint8_t *) &AbsMinHeatSetpointLimit, sizeof(AbsMinHeatSetpointLimit), NULL);
+
+    emberAfReadAttribute(endpoint, ZCL_THERMOSTAT_CLUSTER_ID, ZCL_ABS_MAX_HEAT_SETPOINT_LIMIT_ATTRIBUTE_ID, CLUSTER_MASK_SERVER,
+                         (uint8_t *) &AbsMaxHeatSetpointLimit, sizeof(AbsMaxHeatSetpointLimit), NULL);
+
+    emberAfReadAttribute(endpoint, ZCL_THERMOSTAT_CLUSTER_ID, ZCL_MIN_HEAT_SETPOINT_LIMIT_ATTRIBUTE_ID, CLUSTER_MASK_SERVER,
+                         (uint8_t *) &MinHeatSetpointLimit, sizeof(MinHeatSetpointLimit), NULL);
+
+    emberAfReadAttribute(endpoint, ZCL_THERMOSTAT_CLUSTER_ID, ZCL_MAX_HEAT_SETPOINT_LIMIT_ATTRIBUTE_ID, CLUSTER_MASK_SERVER,
+                         (uint8_t *) &MaxHeatSetpointLimit, sizeof(MaxHeatSetpointLimit), NULL);
+
+    // Make sure the user imposed limits are within the manufacture imposed limits
+
+    if (MinHeatSetpointLimit < AbsMinHeatSetpointLimit)
+        MinHeatSetpointLimit = AbsMinHeatSetpointLimit;
```

Comment:
What does the "If this attribute is not present" verbiage around `MinHeatSetpointLimit` in the spec mean?  How can it not be present if `AbsMinHeatSetpointLimit` is present?

And concretely, can we get an error reading `MinHeatSetpointLimit` but not an error reading `AbsMinHeatSetpointLimit`?

Reply:
Concretely, yes.  Each of the limit attributes are specified as optional with no dependencies.  
So for  a device, the manufacturer could specify their limits, but not provide the ability for a user to make tighter limits.

Reply:
Ah, right they are `[HEAT]`, not `HEAT`.

OK.  Given that, this code seems wrong to me.  Consider the case when `MinHeatSetpointLimit` is not implemented but `AbsMinHeatSetpointLimit` is implemented and its value is below the default 7C value; say it's 5C.

In that situation we will end up setting `MinHeatSetpointLimit` here to 7C because we will use the default value for `MinHeatSetpointLimit` (due to it not being implemented and the read failing etc), and that value is not smaller than `AbsMinHeatSetpointLimit`, so we don't adjust it here.  But the spec clearly says:

> If this attribute is not present, it SHALL be taken as equal to AbsMinHeatSetpointLimit.

so in this case the value should end up 5C, not 7C, right?

Similar for the other min/max bits in this PR.

Reply:
Done - Also addressed the case of Min/Max limits are present and Abs* are not so Min/Max should be accepted. 
Again maybe this is a spec issue.

Reply:
https://github.com/CHIP-Specifications/connectedhomeip-spec/issues/3724